sql-plugin/src/main/scala/com/nvidia/spark/rapids/profiler.scala

```diff
@@ -118,10 +121,34 @@ object ProfilerOnExecutor extends Logging {
       val taskCtx = TaskContext.get
       val stageId = taskCtx.stageId
       if (stageRanges.contains(stageId)) {
-        synchronized {
-          activeStages.add(taskCtx.stageId)
-          enable()
-          startPollingDriver()
+        if (stageTaskLimit <= 0) {
+          // Unlimited tasks per stage
+          synchronized {
+            activeStages.add(taskCtx.stageId)
+            enable()
+            startPollingDriver()
+          }
+        } else {
+          // Limited tasks per stage
+          val currentCount = synchronized(stageTaskCount.getOrElseUpdate(stageId, 0))
+          if (currentCount< stageTaskLimit) {
+            enable()
+            startPollingDriver()
```

Comment:
This should be synchronized as it is in the no-stage-task-limit case so we don't have two threads trying to enable in parallel.